stu.1
.\" Autogenerated on Fri Mar 24 22:23:09 CET 2023 by sh/mkman
.TH STU 1 "March 2023" "stu-2.7.14" "STU"
.SH NAME
stu \- Build automation
.SH SYNOPSIS
.B stu 
[ -f
.I FILENAME
] [
.I OPTION 
]... [
.I TARGET
]...
.SH DESCRIPTION
Stu is a tool for build automation.  Like Make, it is used to call other
programs in 
order to build files.  Stu has three features that work in
conjunction to allow contructs impossible to realize in Make: 
(1) Dynamic dependencies:  The dependencies of a target can be
themselves computed by a command.  When '[FILENAME]' (a filename
currounded by angle brackets) is used as a 
dependency, Stu will build the file FILENAME and read dependencies from it.  
(2) Parameters:  Names of files (and transient targets) can contain
parameters written using dollar syntax.  For example, a rule with a target
filename 'list.$name' will be used to build any file matching the
pattern 'list.*', and the parameter $name can be used in names of
dependencies and in the build command.  Targets may contain any number
of parameters. 
(3) Concatenation:  Stu allows expressions such as list.[FILENAME] and
list.(a b c) to construct a list of files following a certain pattern.

At its core, Stu is very much like Make: Instead of a 'Makefile', one
uses a file named 'main.stu', which is read by Stu. Also like Make, Stu
will build the first target it finds, except if given an explicit target
to built.  Stu is designed to follow the conventions of Make, of the
shell, and of Unix and POSIX in general.  As an example, the options
.BR -f
and
.BR -k
have the same meaning in Stu as they do in Make. 

Rules for building files are read from the file 'main.stu', 
from a file passed by the
.BR -f
option, or using 
.BR -F . 

Names of targets to build passed on the command line outside of options follow a simplified Stu syntax,
in which the only special characters are the transient target marker 
.BR '@' , 
brackets 
.BR '[]' ,
and flags starting with a dash
.BR '-' .
All other characters
are interpreted as part of filenames; this includes all spaces.
The parentheses
.BR '()'
are not parsed as Stu syntax, because the shell's braces
.BR '{}'
can be used instead.  When 
.BR -J 
is used,
arguments passed on the command line are interpreted as filenames
without Stu syntax. 
The syntax of arguments outside of options and without using the 
.BR -J
is subject to change with new features in Stu -- to pass filenames portably, use the 
.BR -c
option when passing individual filenames, or 
.BR -J
when passing multiple filenames, e.g. via 
.BR xargs (1). 
To pass targets in full stu syntax portably, use the 
.BR -C
option. 

.SH OPTIONS
.IP "-0 FILENAME"
Read targets from FILENAME, which must contain 
\\0-separated filenames.
No Stu syntax is processed.  This option is also suitable in all cases
where a file contains only the name of a single file, since filenames do
not contain the \\0 character. 
Using this option is equivalent to using the
.BR "[-0 FILENAME]" 
syntax.
.IP -a
Treat all trivial dependencies, which are declared with the
.BR -t
flag or option, as non-trivial.
.IP "-c FILENAME"
Pass a target filename, without Stu syntax.  This option only allows
file targets to be specified, not transient targets. 
.IP "-C EXPRESSION"
Pass a target using full Stu syntax.  The given expression is parsed as
if it was given as one or more dependencies in a Stu script.  As a result,
flags 
.BR -p , 
.BR -o 
and 
.BR -t , 
brackets, parentheses and other Stu syntax elements
can be used, and spaces separate filenames.  This is not equivalent to
passing a string as an argument to Stu outside of options, because
.BR -C 
supports full Stu syntax, while arguments supports only a reduced syntax
(see above). 
.IP -d
Debug mode.  Show internal information about the Stu algorithm on
standard error output.  All printed output lines begin with the
string 'DEBUG  ', i.e. the word DEBUG followed by two spaces.  The
output is otherwise not standardized and is subject to change, in
particular with internal changes to the algorithms used. 
.IP "-E"
Explain error messages.  For certain errors, an additional explanation is
written on standard error output.  Only some error messages have explanations. 
.IP "-f FILENAME"
Read the given file instead of the default 'main.stu'.  If the name of a
directory is passed, the file 'main.stu' in the given directory is
read.  If '-' is passed, read standard input.  The
.BR -f
option can be
used multiple times.  
.IP "-F RULES"
Pass rules in full Stu syntax.  
.IP -g
Treat all optional dependencies (declared with the
.BR -o
flag) as non-optional.
.IP -h
Output a short help and exit.
.IP "-i"
Interactive mode.  I.e., put the jobs run into the foreground.  Must not
be used in conjunction with
.B -j N
when N > 1.  Without this option, jobs are run in the
background, meaning they can't read standard input (they get /dev/null instead), will not get any
terminal signals, and pressing Ctrl-C will stop Stu as a whole.  Using
this option, each job is run in the foreground, and therefore they can
read standard input, will get terminal signals such as SIGINT (after
Ctrl-C) and end-of-stdin (after Ctrl-D).  Thus, when
.BR -i
is used, pressing Ctrl-C will only interrupt the currently running job,
and if the
.BR -k
option is used, Stu will continue with the next job.  When Ctrl-Z
is pressed, Stu will interrupt the currently running job, and ask the
user to press ENTER to continue, or press Ctrl-C/Ctrl-Z to interrupt or
suspend Stu itself. 
When 
.BR -i
is used, Stu must be run in a terminal. 
.IP "-j K"
Run K jobs in parallel.  K must be a positive integer.  Without this
option, jobs are not run in parallel, which is equivalent to using the 
.BR "-j 1" 
setting.
When running jobs in parallel and the
.BR -k
option is not used, a single job that fails will make Stu abort all
other running jobs and terminate.  Thus, the
.BR -j
option is often used in conjunction with the
.BR -k
option. 
The parameter K is mandatory.
This option works like the corresponding option in GNU Make, but note
that in GNU Make, the argument is optional. 
.IP "-J"
Parse all arguments to Stu as filenames, disabling all Stu syntax that
is otherwise used.  Intended when Stu is used with tools such
as
.BR xargs (1)
or similar.  The
.BR -J
option itself does not take an argument; it only
changes the way arguments outside of options are interpreted.
.IP -k
Keep going.  Do not stop running when an error occurs.  Instead, try to build as much
as possible.  This option is equivalent to that of Make. 
.IP "-K"
Don't delete partially built files when a command fails or when Stu is
interrupted.  By default, Stu will delete the target of a command after
the command fails or is interrupted, when the file is newer than it was
before starting the command. This option disables that behavior.  Note
that with this option, a subsequent invocation of Stu may lead to the
partially built file being erroneously considered up to date. 
.IP "-m ORDER"
Specify the order in which jobs are run.  When ORDER is 'dfs' (the default),
Stu traverses the dependency graph in a depth-first fashion, in a way
similar to most Make implementations. When ORDER is 'random', the order in which jobs are run
is randomized within each target.  
.IP "-M STRING"
Run jobs in pseudorandom order, seeded by the given string. 
.IP "-n FILENAME"
Read targets from FILENAME, which must contain newline-separated
filenames.  No Stu syntax is processed.  Using this option is equivalent to using the
.BR "[-n FILENAME]" 
syntax.
.IP "-o FILENAME"
Pass the given file as an optional dependency, i.e., build it only if it
already exists and is out of date. 
.IP "-p FILENAME"
Pass the given file as a persistent dependency, i.e., build the file but
ignore its timestamp. 
.IP "-P"
Print the set of rules to standard output and exit.  Note that Make 
has a
.BR -p
option to print the rules, but does not exit.  The output
format is not specified. 
.IP "-q"
Question mode.  Do not execute any commands.  Instead, determine whether
the targets are up to date.  The exit status is 0 when all given
targets (or the default target) are up to date, and 1 when not.  The
exit status may still be 2 or 4 on encountering logical or fatal
errors. 
The options 
.BR -k
and 
.BR -j 
are ignored.
.IP "-s"
Silent mode.  Suppress messages on standard output:  messages about
which commands are run, a message when the build is successful, and a
message when there is nothing to be done.  Error messages are not
suppressed.  This option is comparable to the same option in Make.  
.IP -V 
Output the version number of Stu and exit.
.IP "-x"
Call the shell using the
.BR -x
option, i.e., each individual shell command is output to standard error
output individually, instead of 
outputting a full command at once on standard output.  In the output,
each command is prefixed by the value of '$PS4'. 
.IP -y
Disable color in output.  By default, Stu checks whether error output
and standard error output are TTYs and whether $TERM is defined and
not 'dumb' and if they are, uses ANSI escape sequences to color code
messages.   
.IP -Y
Enable color output unconditionally. 
.IP -z 
Output runtime statistics about child processes on standard output when
finished.  Does not include the runtime of the Stu process itself.
Includes the runtime of all child and grandchild processes, and so on.
Does not include the runtime of children or grandchildren that have not
been waited for (which only happens when Stu is interrupted by a
signal.) 

Stu options are parsed with
.BR getopt(3)
and therefore options must precede arguments.  Options following
arguments may be supported on some platforms.  

.SH OVERVIEW
A simple rule looks as follows:

    results.txt:  data.txt compute {
        ./compute --input data.txt --output results.txt 
    }

The colon may be omitted when there are no dependencies:

    data.csv { ./generate --output data.csv }

Here is an example of a rule containing three parameters.  Stu will use
pattern matching to match the target pattern to a given filename: 

    plot.$dataset.$method.$measure.eps:  
        data-$dataset.txt analyse-$method 
    {
        ./analyse-$method \\
            -m $measure \\
            -f data-$dataset.txt \\
            -o plot.$dataset.$method.$measure.eps
    }

Here is an example of a dynamic dependency.  The target 'compute' (a C
program) must be rebuild whenever its source code files are modified.
Since the set of source code files is large and may be changed by
changing the source code itself, we use the file 'compute.c.dep' to
contain the list of dependencies.  The file 'compute.c.dep' will then be
built by Stu like any file, and its content parsed for the actual
dependencies:  

    compute:  [compute.c.dep] {
        gcc -c compute.c -o compute 
    }
    $name.c.dep:  $name.c compute-dep {
        ./compute-dep-c "$name.c" >"$name.c.dep"
    }

Parameters can also use the syntax ${...}.

Syntax can be on multiple lines; whitespace is not significant.  No
backslashes are needed at line ends:

    output.txt: 
        a.data b.data c.data d.data e.data f.data g.data h.data i.data
        j.data k.data l.data m.data
    {
        do-stuff >output.txt; 
    }

A rule may be entirely given on a single line: 

    system-info: { uname -a >system-info }

The following rule uses single quotes to declare filenames that include
parentheses and a colon:  

    '((':  'aaa:bbb' {
        ./bla -f 
    }

Multiple parametrized rules may match a target.  In that case Stu uses
the one that is the least parametrized, as defined by the subset
relation on the set of characters that are in parameters. 
When building 'X.txt' in this example, only the second rule is called:

    $name.txt: {  echo "$name" is the best >"$name.txt" }
    X.txt:  { echo X sucks >X.txt }

Persistent dependencies:  In the following example, the
directory 'data' is a persistent dependency, i.e. 'data' is only
built when it does not exist, but it is never re-built.  A
persistent dependency is indicated by the 
.BR -p 
flag.  This is
useful for directories, whose timestamps change when files are
created/removed in them.  

    data/file:  -p data {
        echo Hello >data/file
    }
    data: { mkdir data }

Optional dependencies can be declared with the 
.BR -o 
flag.  An optional
dependency will never be built if it does not already exist.  If it
already exists, then its own rule is used (and its date checked) to
decide whether it should be rebuilt.  

    target:  -o input {
        if [ -r input ] ; then
            cp input target
        else
            echo Hello >target
        fi
    }

Trivial dependencies are denoted with the 
.BR -t 
flag.  They denote a
dependency that should never cause a target to be rebuilt, but if the
target is rebuilt for another reason, then they are treated like normal
dependencies.  Trivial dependencies don't even cause a target to be
rebuilt if they don't exist.  
Trivial dependencies are typically used for
configuration, i.e., for the setting up configuration of application.
Trivial dependencies are not allowed if the rule has no command. 

    target:  -t input { ... }

Variable dependency:  the content of variables can come from files.  
In the following example, the C flags are stored in the file 'CFLAGS',
and used in the compilation command using the $[CFLAGS] dependency.  

    compute:  compute.c $[CFLAGS]
    {
        gcc $CFLAGS compute -o compute.c
    }
    CFLAGS: { echo -Wall -Werror >CFLAGS }

Variable dependencies may be declared as persistent as in 
.B $[-p X]
and as trivial as in 
.B $[-t X]
but not as optional using the 
.B -o 
flag. 
By default, the name of the variable set is the same as the filename.
Another variable name can be used in the following way:

    $[NAME = FILENAME]

If multiple variable dependencies have the same name, it is unspecified
which one is used.  If a variable dependencies has the same name as a
parameter, it overrides the parameter. 

Transient targets are marked with '@'.  They are used for targets such
as '@clean' that do an action without building a file, and for lists of
files that depend on other targets, but don't have a command associated

NEWS
Stu versions are of the form X.Y.Z, where X is the major version, Y is
the minor version and Z the patch level.  For each new patch level, no
new syntax features are added.  When the minor version is incremented,
features are added, which are documented in this file.  This file also
documents important changes in patch level updates, which are mostly
changes in terminology, changes in the installation procedure, or
changes in invocation. 

When the major version is increased, backward-incompatible changes are
made, which are documented in the file 'MIGRATION', and summarized here. 

------------------------------------------------------------------------

2020-10-30  Version 2.7

* Backslashes are now allowed outside of quotes and act as the escape character: 
  * In front of newline, the newline is ignored
  * At end of file:  An error
  * Before any other character:  Represents the character as part of a unquoted
    filename. 

2020-04-13  Version 2.6.4

* Switched to C++14 (from C++11). 

2020-03-18  Version 2.6

* Canonicalization is implemented.  This means that Stu will now
  canonicalize all names of files and transient before further
  processing.  For instance, 'data' and './data' will refer to the same
  file.  This takes into account uses of '/' and '.'.  Symlinks and '..'
  are not resolved by canonicalization.  The exact semantics are described in
  the manpage. 
  

2018-02-28  Version 2.5.59

* We clarified what happens when multiple environment variables that Stu
  sets have identical names:  
  (1) Two parameter cannot have the same name; they would be interpreted
  as the same parameter.
  (2) Two variable dependencies may have the same name.  This is
  allowed, and it is unspecified which gets used.  
  (3) If a variable dependency and a parameter have the same name, the
  variable overrides the parameter. 

2018-02-15  Version 2.5.53

* The parsing of arguments on the command line was made consistent with
  the new features introduced in Version 2.5.  As before, the
  characters @[] are recognized, and now also the '-' character for
  flags.  Empty arguments now result in logical error (exit status 2)
  rather than fatal errors (exit status 4). 

2017-08-11  Version 2.5

* Concatenation is supported
* The handling of dynamic dependencies was rewritten to allow
  concatenation.  The semantics of existing functionality was not
  changed, except in certain undocumented corner cases.  These include
  transients whose rules contain a command or file targets.  These acted
  transitively previously, but do not now.  Only transient rules without
  command and without file targets act transitively now.  This applies
  to transients used in dynamic dependencies, transients used to import
  variable dependencies, as well as transients used to carry over flags
  to child targets. 
* Arguments to -j cannot be in octal anymore.  In previous versions, the
  -j option accepted numbers in octal when they started with a zero.
  This behavior was accidental, and has been removed. 
* When a file read using the -f option or a file included using %include
  does not exist, Stu now returns an exit code of 1 (build error),
  rather than 2 (logical error) as done previously. 

2016-09-13  Version 2.4.19

* The rules governing what is written on stdout where reworked:  There
  is now only the -s option (silent).  The options -Q and -w were
  removed. 

2016-09-11  Version 2.4.17

* "Verbose mode" is now called "debug mode", and the option was changed
  from -v to -d.  

2016-09-03  Version 2.4.7

* The -B option is renamed to -i, because it is mainly an interactive
  mode, even if only very rudimentary. 

2016-08-31  Version 2.4.2

* Job control:  Jobs are now always put into the background, except when
  the -B function is used.  The -B function must now be used when Stu
  command should read from their standard input. 

2016-08-31  Version 2.4

* Copy rules can now use the -o flag to denote optional copy rules:

A = -o B;

The semantics are:
- When B is not present:
   - Fail when both A is not present
   - Succeed when A is present 
- When B is present, apply the usual rules 

This is particularly useful when working in projects where a certain file
can be updated from a source only for some developers:  The file can be
distributed for the project, and an optional copy rule can be used such
that for developers who have the source file, it will be updated when
necessary, and for developers who don't have the source file, the copy
delivered with the project can always be used. 

2016-08-31  Version 2.3.10

* The environment variable $STU_OPTIONS can now be used to pass options
  to Stu.  All non-option charaters are ignored, i.e., you can include
  spaces an preceded all options with a dash, but it doesn't matter.
  This works only for options for which it makes sense to set them
  globally:  "EQswxyYz".  Options passed on the command line take
  precedence over options passed in $STU_OPTIONS. 

2016-08-30  Version 2.3.6

* Stu now starts all commands with their standard input redirected from
  /dev/null, except for command whose rule has an input redirection.
  This behavior, along with the usage of a separate process group for
  each job, can be disabled with the '-B' option.  Thus, jobs will now
  systematically get a STDIN of size zero instead of (erratically)
  getting Stu's STDIN.  If jobs should be able to read Stu's STDIN, then
  '-B' has to be used.  For technical reasons, this cannot be used in
  conjunction with process groups, and as a result, it may happen that
  indirect child processes (i.e., grandchild processes) may not be
  terminated correctly by Stu when -B is used. 

2016-08-29  Version 2.3

* A flag and an option '-0' is now available for \0-separated dynamic
  dependencies, that work analogously to '-n' for newline-separated
  dependency.  Thus, this is now possible (using the nonportable find
  -print0 as an example):

@all [ -0 FILES ];
>FILES { find ... -print0 }


src/stu.cc
/* 
 * This is the top-level source code file, which contains the main()
 * function.  See the manpage for a description of options, the exit
 * status, etc.
 */

/* Enable bounds checking when using GNU libc.  Must be defined before
 * including any of the standard headers.  (Only in non-debug mode).  A
 * no-op for non-GNU libc++ libraries.  */ 
#ifndef NDEBUG
#    define _GLIBCXX_DEBUGa
#endif 

#include <unistd.h>
#include <sys/time.h>

#include <memory>
#include <vector>

/* Used for all of Stu */
using namespace std; 

#include "dep.hh"
#include "execution.hh" 
#include "rule.hh"
#include "timestamp.hh"
#include "color.hh"

/*
 * Note:  Stu does not call setlocale(), and therefore can make use of
 * isspace() detecting spaces as defined in the C locale.  The same is
 * true for isalnum(), which is used to get ASCII results. 
 */

/* 
 * We use getopt(), which means that Stu does only support short
 * options, and not long options.  We avoid getopt_long() as it is a GNU
 * extension, and the short options are sufficient for now. 
 *
 * Also, using getopt() means that the exact synytax of Stu depends on
 * the platform:  GNU getopt() will all options to follow arguments,
 * while BSD getopt() does not. 
 */
const char OPTIONS[]= "0:ac:C:dEf:F:ghij:JkKm:M:n:o:p:PqsVxyYz"; 

/* The output of the help (-h) option.  The following strings do not
 * contain tabs, but only space characters.  */   
const char HELP[]= 
	"Usage: " PACKAGE " [-f FILENAME] [OPTION]... [TARGET]...\n"           
	"By default, build the first target in the file 'main.stu'.\n" 
	"TARGET may include the special characters '@[]-'.\n"     
	"Options:\n"						       
	"  -0 FILENAME      Read \\0-separated file targets from the given file\n"
	"  -a               Treat all trivial dependencies as non-trivial\n"          
	"  -c FILENAME      Pass a target filename without Stu syntax parsing\n"      
	"  -C EXPRESSIONS   Pass a target in full Stu syntax\n"		              
	"  -d               Debug mode: show execution information on stderr\n"     
	"  -E               Explain error messages\n"                                 
	"  -f FILENAME      The input file to use instead of 'main.stu'\n"            
	"  -F RULES         Pass rules in Stu syntax\n"                               
	"  -g               Treat all optional dependencies as non-optional\n"        
	"  -h               Output help and exit\n"		                      
	"  -i               Interactive mode (run jobs in foreground)\n"
	"  -j K             Run K jobs in parallel\n"			              
	"  -J               Disable Stu syntax in arguments\n"                        
	"  -k               Keep on running after errors\n"		              
	"  -K               Don't delete target files on error or interruption\n"     
	"  -m ORDER         Order to run the targets:\n"			      
	"     dfs           (default) Depth-first order, like in Make\n"	      
	"     random        Random order\n"				              
	"  -M STRING        Pseudorandom run order, seeded by given string\n"         
	"  -n FILENAME      Read \\n-separated file targets from the given file\n"
	"  -o FILENAME      Build an optional dependency, i.e., build it only if it\n"
	"                   exists and is out of date\n"
	"  -p FILENAME      Build a persistent dependency, i.e., ignore its timestamp\n"
	"  -P               Print the rules and exit\n"                               
	"  -q               Question mode: check whether targets are up to date\n"    
	"  -s               Silent mode: don't use stdout\n"
	"  -V               Output version and exit\n"				      
	"  -x               Output each line in a command individually\n"              
	"  -y               Disable color in output\n"                                
	"  -Y               Enable color in output\n"
	"  -z               Output run-time statistics on stdout\n"                   
	"Report bugs to: " PACKAGE_BUGREPORT "\n" 
	"Stu home page: <" PACKAGE_URL ">\n";

const char VERSION_INFO[]=
	PACKAGE " " STU_VERSION "\n"
	"Copyright (C) 2014-2023 Jerome Kunegis\n"
	"License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
	"This is free software: you are free to change and redistribute it.\n"
	"There is NO WARRANTY, to the extent permitted by law.\n";

void init_buf(); 
/* Initialize buffers; called once from main() */ 

void add_deps_option_C(vector <shared_ptr <const Dep> > &deps,
		       const char *string_);
/* Parse a string of dependencies and add them to the vector. Used for
 * the -C option.  Support the full Stu syntax.  */

/* Set one of the "setting options", i.e., of of those that can appear
 * in $STU_OPTIONS.  Return whether this was a valid settings option.  */ 
bool stu_setting(char c)
{
	switch (c) {
	default:  return false;

	case 'E':  option_explain= true;        break;
	case 's':  option_silent= true;         break;
	case 'x':  option_individual= true;     break;
	case 'y':  Color::set(false);           break;
	case 'Y':  Color::set(true);            break;
	case 'z':  option_statistics= true;     break;
	}

	return true; 
}

int main(int argc, char **argv, char **envp)
{
	/* Initialization */
	dollar_zero= argv[0]; 
	envp_global= (const char **) envp; 
	init_buf();
	Job::init_tty();
	Color::set();
	int error= 0;

	/* Refuse to run when $STU_STATUS is set */ 
	const char *const stu_status= getenv("STU_STATUS");
	if (stu_status != nullptr) {
		print_error(frmt("Refusing to run recursive Stu; "
				 "unset %s$STU_STATUS%s to circumvent",
				 Color::word, Color::end));
		exit(ERROR_FATAL); 
	}

	try {
		vector <string> filenames;
		/* Filenames passed using the -f option.  Entries are
		 * unique and sorted as they were given, except for
		 * duplicates. */   

		vector <shared_ptr <const Dep> > deps; 
		/* Assemble targets here */ 

		shared_ptr <const Place_Param_Target> target_first; 
		/* Set to the first rule when there is one */ 

		Place place_first;
		/* Place of first file when no rule is contained */ 

		bool had_option_target= false;   
		/* Whether any target(s) was passed through one of the
		 * options -c, -C, -o, -p, -n, -0.  Also set when zero
		 * targets are passed through one of these, e.g., when
		 * -n is used on an empty file.  */

		bool had_option_f= false; /* Both lower and upper case */

		/* Parse $STU_OPTIONS */ 
		const char *stu_options= getenv("STU_OPTIONS");
		if (stu_options != NULL) {
			while (*stu_options) {
				char c= *stu_options++;
				if (c == '-' || isspace(c))
					continue; 
				if (! stu_setting(c)) {
					Place place(Place::Type::ENV_OPTIONS);
					place << fmt("invalid option %s",
						     multichar_format_err(frmt("-%c", c)));
					exit(ERROR_FATAL); 
				}
			}
		}
		
		for (int c; (c= getopt(argc, argv, OPTIONS)) != -1;) {

			if (stu_setting(c))
				continue;

			switch (c) {

			case 'a': option_nontrivial= true;     break;
			case 'd': option_debug= true;          break;
			case 'g': option_nonoptional= true;    break;
			case 'h': fputs(HELP, stdout);         exit(0);
			case 'J': option_literal= true;        break;
			case 'k': option_keep_going= true;     break;
			case 'K': option_no_delete= true;      break;
			case 'P': option_print= true;          break;  
			case 'q': option_question= true;       break;

			case 'c':  {
				had_option_target= true; 
				Place place(Place::Type::OPTION, 'c');
				if (*optarg == '\0') {
					place << "expected a non-empty argument"; 
					exit(ERROR_FATAL);
				}
				deps.push_back
					(make_shared <Plain_Dep>
					 (0, Place_Param_Target
					  (0, Place_Name(optarg, place))));
				break;
			}

			case 'C':  {
				had_option_target= true; 
				add_deps_option_C(deps, optarg);
				break;
			}

			case 'f':
				if (*optarg == '\0') {
					Place(Place::Type::OPTION, 'f') <<
						"expected non-empty argument"; 
					exit(ERROR_FATAL);
				}

				for (string &filename:  filenames) {
					/* Silently ignore duplicate input file on command line */
					if (filename == optarg)  goto end;
				}
				had_option_f= true;
				filenames.push_back(optarg); 
				Parser::get_file(optarg, -1, Execution::rule_set,
						 target_first, place_first);
			end:
				break;

			case 'F':
				had_option_f= true;
				Parser::get_string(optarg, Execution::rule_set, target_first);
				break;

			case 'i':
				option_interactive= true;
				if (Job::get_tty() < 0) {
					Place place(Place::Type::OPTION, 'i');
					print_warning(place, "Interactive mode cannot be used because no TTY is available"); 
				}
				break;
		
			case 'j':  {
				errno= 0;
				char *endptr;
				Execution::jobs= strtol(optarg, &endptr, 10);
				Place place(Place::Type::OPTION, c); 
				if (errno != 0 || *endptr != '\0') {
					place << fmt("expected the number of jobs, not %s",
						     name_format_err(optarg)); 
					exit(ERROR_FATAL); 
				}
				if (Execution::jobs < 1) {
					place << fmt("expected a positive number of jobs, not %s",
						     name_format_err(optarg));
					exit(ERROR_FATAL); 
				}
				option_parallel= Execution::jobs > 1; 
				break;
			}

			case 'm':
				if (!strcmp(optarg, "random"))  {
					order= Order::RANDOM;
					/* Use gettimeofday() instead of time()
					 * to get millisecond instead of second
					 * precision */ 
					struct timeval tv;
					if (gettimeofday(&tv, nullptr) != 0) {
						print_error_system("gettimeofday");
						exit(ERROR_FATAL); 
					}
					buffer_generator.seed(tv.tv_sec + tv.tv_usec); 
				}
				else if (!strcmp(optarg, "dfs"))     /* Default */ ;
				else {
					print_error(fmt("Invalid argument %s for option %s-m%s; valid values are %s and %s", 
							name_format_err(optarg),
							Color::word, Color::end,
							name_format_err("random"),
							name_format_err("dfs"))); 
					exit(ERROR_FATAL); 
				}
				break;

			case 'M':
				order= Order::RANDOM;
				buffer_generator.seed(hash <string> ()(string(optarg)));
				break;

			case 'n':
			case '0':  {
				had_option_target= true; 
				Place place(Place::Type::OPTION, c); 
				if (*optarg == '\0') {
					place << "expected a non-empty argument";
					exit(ERROR_FATAL);
				}
				deps.push_back
					(make_shared <Dynamic_Dep>
					 (0, make_shared <Plain_Dep>
					  (1 << flag_get_index(c), 
					   Place_Param_Target
					   (0, Place_Name(optarg, place)))));
				break;
			}

			case 'o':
			case 'p':  {
				had_option_target= true; 
				Place place(Place::Type::OPTION, c);
				if (*optarg == '\0') {
					place << "expected a non-empty argument";
					exit(ERROR_FATAL);
				}
				Place places[C_PLACED];
				places[c == 'p' ? I_PERSISTENT : I_OPTIONAL]= place; 
				deps.push_back
					(make_shared <Plain_Dep>
					 (c == 'p' ? F_PERSISTENT : F_OPTIONAL, places,
					  Place_Param_Target(0, Place_Name(optarg, place))));
				break; 
			}

			case 'V': 
				fputs(VERSION_INFO, stdout); 
				printf("USE_MTIM = %u\n", USE_MTIM); 
				exit(0);

			default:  
				/* Invalid option -- an error message was
				 * already printed by getopt() */   
				string text= name_format_err(frmt("%s -h", dollar_zero));
				fprintf(stderr, 
					"To get a list of all options, use %s\n", 
					text.c_str()); 
				exit(ERROR_FATAL); 
			}
		}

		order_vec= (order == Order::RANDOM);

		if (option_interactive && option_parallel) {
			Place(Place::Type::OPTION, 'i')
				<< fmt("parallel mode using %s cannot be used in interactive mode",
				       multichar_format_err("-j")); 
			exit(ERROR_FATAL); 
		}

		/* Targets passed on the command line, outside of options */ 
		for (int i= optind;  i < argc;  ++i) {
			/* The number I may not be the index that the
			 * argument had originally, because getopt() may
			 * reorder its arguments. (I.e., using GNU
			 * getopt.)  This is why we can't put I into the
			 * trace.  */ 

			Place place(Place::Type::ARGUMENT); 
			if (*argv[i] == '\0') {
				place << fmt("%s: name must not be empty",
					     name_format_err(argv[i])); 
				if (! option_keep_going) 
					throw ERROR_LOGICAL;
				error |= ERROR_LOGICAL; 
			} else if (option_literal)
				deps.push_back(make_shared <Plain_Dep> 
					       (0, Place_Param_Target
						(0, Place_Name(argv[i], place))));
		}

		if (! option_literal) {
			Parser::get_target_arg(deps, argc - optind, argv + optind); 
		} 

		/* Use the default Stu script if -f/-F are not used */ 
		if (! had_option_f) {
			filenames.push_back(FILENAME_INPUT_DEFAULT); 
			int file_fd= open(FILENAME_INPUT_DEFAULT, O_RDONLY); 
			if (file_fd >= 0) {
				Parser::get_file("", file_fd, 
						 Execution::rule_set, target_first,
						 place_first); 
			} else {
				if (errno == ENOENT) { 
					/* The default file does not exist --
					 * fail if no target is given */  
					if (deps.empty() && ! had_option_target 
					    && ! option_print) {
						print_error(fmt("Expected a target or the default file %s",
								name_format_err(FILENAME_INPUT_DEFAULT))); 

						explain_no_target(); 
						throw ERROR_LOGICAL; 
					}
				} else { 
					/* Other errors by open() are fatal */ 
					print_error_system(FILENAME_INPUT_DEFAULT);
					exit(ERROR_FATAL);
				}
			}
		}

		if (option_print) {
			Execution::rule_set.print(); 
			exit(0); 
		}

		/* If no targets are given on the command line,
		 * use the first non-variable target */ 
		if (deps.empty() && ! had_option_target) {

			if (target_first == nullptr) {
				if (! place_first.empty()) {
					place_first
						<< "expected a rule, because no target is given";
				} else {
					print_error("No rules and no targets given");
				}
				exit(ERROR_FATAL);
			}

			if (target_first->place_name.is_parametrized()) {
				target_first->place <<
					fmt("the first target %s must not be parametrized if no target is given",
					    target_first->format_err());
				exit(ERROR_FATAL);
			}

			deps.push_back(make_shared <Plain_Dep> (*target_first));  
		}

		/* Execute */
		Execution::main(deps);

	} catch (int e) {
		assert(e >= 1 && e <= 3); 
		error= e;
	}

	/*
	 * Code executed before exiting:  This must be executed even if
	 * Stu fails (but not for fatal errors).
	 */
	
	if (option_statistics) {
		Job::print_statistics();
	}

	if (fclose(stdout)) {
		perror("fclose(stdout)");
		exit(ERROR_FATAL);
	}
	/* No need to flush stderr, because it is line buffered, and if
	 * we used it, it means there was an error anyway, so we're not
	 * losing any information  */

	exit(error); 
}

void init_buf()
{
	/* Set STDOUT to line buffered, so that the
	 * output of command lines always happens before the output of
	 * commands themselves */ 
	/* Note:  Setting the buffering like this is only possible if we
	 * have not written anything yet.  */  
	if (0 != setvbuf(stdout, nullptr, _IOLBF, 0)) {
		print_error_system("setvbuf"); 
		exit(ERROR_FATAL); 
	}
		
	/* Set STDOUT to append mode; this is also done by GNU Make */ 
	{
		int flags= fcntl(fileno(stdout), F_GETFL, 0);
		if (flags >= 0)
			fcntl(fileno(stdout), F_SETFL, flags | O_APPEND);
	}

	/* Set STDERR to append mode; this is also done by GNU Make */ 
	{
		int flags= fcntl(fileno(stderr), F_GETFL, 0);
		if (flags >= 0)
			fcntl(fileno(stderr), F_SETFL, flags | O_APPEND);
	}
}

void add_deps_option_C(vector <shared_ptr <const Dep> > &deps,
		       const char *string_)
{
	vector <shared_ptr <Token> > tokens;
	Place place_end;
				
	Tokenizer::parse_tokens_string
		(tokens, 
		 Tokenizer::OPTION_C,
		 place_end, string_,
		 Place(Place::Type::OPTION, 'C'));

	vector <shared_ptr <const Dep> > deps_option;
	Place_Name input; /* remains empty */ 
	Place place_input; /* remains empty */ 

	Parser::get_expression_list(deps_option, tokens, 
				    place_end, input, place_input);

	for (auto &j:  deps_option) {
		deps.push_back(j); 
	}
}


